### DISCUSSION TITLE
How does cp-editor use competitive companion?

### DISCUSSION THREAD

**Discussion by tusqasi** [cpeditor/cpeditor#807] [Q&A] [answered]
I want to use competitive companion in a program I am making. I can't make a good solution for it. 

Can anyone tell me how does cp-editor uses competitive companion internally?

Competitive companion sends a post request, but how does cp-editor listen for that request.

**Comment by coder3101** [comment, marked as answer]
As you mentioned cpeditor just listens for our port `10045` which competitive companions already knows and sends data to. As soon as you enable competitive companion (enabled by default), we start an http server and start listening for connections on that port. We do some sanitation like the request type and payload verification, then we parse it and send to editor to open as new tab.

These lines create and parses the data sent on that port.
https://github.com/cpeditor/cpeditor/blob/2873c5a6f6f66d8f44fcb9d03bd5515a88def3f5/src/Extensions/CompanionServer.cpp#L56-L96

We use [qhttp](https://github.com/cpeditor/qhttp) library for creating the server.

**Reply by tusqasi** [reply]
That was exactly what I wanted to know.

Thank you.
 
This is going to ease my work quite a lot.
